File: support_tools/cpp/upd_commander_checker/src/responsibility_rules.cpp
#include "responsibility_rules.hpp"

#include <regex>
#include <sstream>
#include <string>
#include <vector>

namespace upd_checker {
namespace {

constexpr int kMaxResponsibilityLines = 250;

bool is_blank(const std::string& line) {
    return line.find_first_not_of(" \t\r\n") == std::string::npos;
}

}  // namespace

std::vector<Finding> check_responsibilities(
    const std::string& source_text,
    const std::string& path,
    const std::vector<IgnoreRule>& rules) {
    std::vector<Finding> findings;
    std::istringstream stream(source_text);
    std::string line;
    int line_number = 0;
    int non_blank_lines = 0;
    int major_class_count = 0;
    int second_class_line = 0;
    const std::regex class_pattern(R"(^\s*class\s+[A-Za-z_][A-Za-z0-9_]*\s*(?:[:{]))");

    while (std::getline(stream, line)) {
        ++line_number;
        if (!is_blank(line)) {
            ++non_blank_lines;
        }
        if (std::regex_search(line, class_pattern)) {
            ++major_class_count;
            if (major_class_count == 2) {
                second_class_line = line_number;
            }
        }
    }

    if (non_blank_lines > kMaxResponsibilityLines &&
        !is_ignored(path, "UPD401", "", rules)) {
        findings.push_back(Finding{
            path,
            1,
            "UPD401",
            "file/module is too large for one responsibility",
            "warning",
        });
    }

    if (major_class_count > 1 &&
        !is_ignored(path, "UPD402", "", rules)) {
        findings.push_back(Finding{
            path,
            second_class_line,
            "UPD402",
            "file contains multiple major classes",
            "warning",
        });
    }

    return findings;
}

}  // namespace upd_checker

File: support_tools/cs/upd_commander_checker/ResponsibilityRules.cs
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace UpdCommanderChecker;

internal static class ResponsibilityRules
{
    private const int MaxResponsibilityLines = 250;
    private const int MaxResponsibilityMethods = 12;

    internal static List<Finding> Check(ResponsibilityCheckInput input)
    {
        var findings = new List<Finding>();
        var types = input
            .Root.DescendantNodes()
            .OfType<TypeDeclarationSyntax>()
            .Where(type => type is not InterfaceDeclarationSyntax)
            .ToList();
        var majorTypes = types.Where(HasBehavior).ToList();

        foreach (var type in types)
        {
            var span = type.GetLocation().GetLineSpan();
            var lineCount = span.EndLinePosition.Line - span.StartLinePosition.Line + 1;
            var methodCount = type.Members.OfType<MethodDeclarationSyntax>().Count();
            if (lineCount <= MaxResponsibilityLines && methodCount <= MaxResponsibilityMethods)
            {
                continue;
            }
            if (
                IgnoreRules.IsIgnored(
                    new IgnoreCheckInput(input.Path, "UPD401", string.Empty, input.IgnoreRules)
                )
            )
            {
                continue;
            }

            findings.Add(
                new Finding(
                    input.Path,
                    span.StartLinePosition.Line + 1,
                    "UPD401",
                    $"type {type.Identifier.ValueText} is too large for one responsibility "
                        + $"(lines={lineCount}, methods={methodCount})",
                    "warning"
                )
            );
        }

        if (
            types.Count == 0
            && input.Lines.Count > MaxResponsibilityLines
            && !IgnoreRules.IsIgnored(
                new IgnoreCheckInput(input.Path, "UPD401", string.Empty, input.IgnoreRules)
            )
        )
        {
            findings.Add(
                new Finding(
                    input.Path,
                    1,
                    "UPD401",
                    "file/module approximation is too large for one responsibility",
                    "warning"
                )
            );
        }

        if (
            majorTypes.Count > 1
            && !IgnoreRules.IsIgnored(
                new IgnoreCheckInput(input.Path, "UPD402", string.Empty, input.IgnoreRules)
            )
        )
        {
            var second = majorTypes[1];
            findings.Add(
                new Finding(
                    input.Path,
                    second.GetLocation().GetLineSpan().StartLinePosition.Line + 1,
                    "UPD402",
                    "file contains multiple responsibility-bearing types",
                    "warning"
                )
            );
        }

        return findings;
    }

    private static bool HasBehavior(TypeDeclarationSyntax type)
    {
        return type.Members.Any(MemberHasBehavior);
    }

    private static bool MemberHasBehavior(MemberDeclarationSyntax member)
    {
        if (
            member
            is MethodDeclarationSyntax
                or ConstructorDeclarationSyntax
                or DestructorDeclarationSyntax
                or OperatorDeclarationSyntax
                or ConversionOperatorDeclarationSyntax
                or IndexerDeclarationSyntax
                or EventDeclarationSyntax
        )
        {
            return true;
        }
        if (member is PropertyDeclarationSyntax property)
        {
            return property.ExpressionBody is not null
                || property.AccessorList?.Accessors.Any(accessor =>
                    accessor.Body is not null || accessor.ExpressionBody is not null
                ) == true;
        }
        return false;
    }
}
